exporter/exporterhelper/queue.go

```diff
@@ -0,0 +1,104 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package exporterhelper // import "go.opentelemetry.io/collector/exporter/exporterhelper"
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+)
+
+// QueueConfig defines configuration for queueing requests before exporting.
+// It's supposed to be used with the new exporter helpers New[Traces|Metrics|Logs]RequestExporter.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+// TODO: Consider moving it to config/configqueue module.
+type QueueConfig struct {
+	// Enabled indicates whether to not enqueue batches before exporting.
+	Enabled bool `mapstructure:"enabled"`
+	// NumConsumers is the number of consumers from the queue.
+	NumConsumers int `mapstructure:"num_consumers"`
+	// QueueSizeItems is the maximum number of items (spans, metric data points or log records)
+	// allowed in queue at any given time.
+	QueueSizeItems int `mapstructure:"queue_size_items"`
+	// QueueSize is the maximum number of requests allowed in queue at any given time.
+	// This option is left for backward compatibility and will be deprecated in the future.
+	// It's recommended to use QueueSizeItems instead.
+	QueueSize int `mapstructure:"queue_size"`
```

Comment:
Should we offer 2 sizes or a size + enum saying count num requests or count num items eventually count bytes if we want?

Reply:
I was thinking about it. From the user's perspective, specifying the enum field might be less obvious. Also, setting `queue_element: item` (assuming we find a better field name) where the item represents a span or metric data point seems more confusing than having a set of predefined documented fields. Let me know if you have any other specific suggestions. For now, I think `queue_size`, `queue_size_items`, and potentially `queue_size_mib` is an easier to use interface.

Reply:
maybe use suffix? `queue_size: 10[B|KB|MB..]` (applies to size in mb) then `queue_size: 10[requests]?` (requests) and `queue_size: 10items`

Reply:
I'm still hesitant that it'll make the user experience easier with `items` being ambiguous.

Having it as a part of the field names gives some flexibility to the component implementations, like they can have different name for the `items` field where it's not directly mapped to spans, datapoint log records. Having the parsing logic hardcoded wouldn't allow that. Also, I can imagine a use case when only one type would be supported by a particular exporter. With different fields, it's possible to restrict the options on the exporter side.

Reply:
If we agree to keep different fields, maybe we can also rename them by removing `queue_` prefix since it's supposed be under `sending_queue` group anyway. Can be 
- `max_size_items` and `max_size_mib`
- `size_limit_items` and `size_limit_mib`
- `items_limit` and `mib_limit`

Reply:
@open-telemetry/collector-approvers, can you please share your thoughts on the user interface? The ask is to provide a way to configure queue size in different measures:
1. Number of basic items: log records, spans, data points - this should be the default measure going forward to be consistent with config for batching
2. Number of incoming requests - this is the current measure set in `queue_size`. We likely should keep it as an option even though it's not required.
3. Size in MiB - will be nice to have in the future.

**Options:**

A. Provide different fields for each option
- `max_size_items`, `max_size_requests` and `max_size_mib`
- `size_limit_items`, `size_limit_requests` and `size_limit_mib`
- `items_limit`, `requests_limit` and `mib_limit`

B. Keep `queue_size` and introduce another field for the `queue_measure` enum which can be `items`, `requests` or `MiB`.

C. Support suffixes like `10requests`, `10items` `10[B|KB|MB..]` and keep the `queue_size` field. 

Both _B_ and _C_ would require a plan on how to migrate from the default measure from `requests` to `items`.

Feel free to suggest other ideas or names for the fields.

Reply:
Personally I like in this order (from most liked to least like: C > B > A.

> Both B and C would require a plan on how to migrate from the default measure from requests to items without affecting configs applied by users.

Why would you want to change the default measure?

Reply:
> Main reasons (though we can also validate to not set multiple values) it is confusing if users configure multiple sizes which one to be used, and trying to implement multiple sizes is out of the scope and an overkill.

Right, I wanted to validate that only one field is specified.

> Why would you want to change the default measure?

Because the batch processor is configured in items, I think it should be the same default for the batcher sender. So, it may be confusing to have different measures by default in the ` sending_queue` and the `batching` options.

Reply:
I removed the `queue_size_items` and left only the backward compatible `queue_size` until we agree on the user config and a migration path. I'll create another issue to move this discussion

Reply:
Submitted https://github.com/open-telemetry/opentelemetry-collector/issues/9462